### ISSUE TITLE
Add new commands to support uploading and downloading images

### ISSUE THREAD

**Issue by qweeah** [oras-project/oras#1366]
### What is the version of your ORAS CLI

oras v1.1.0

### What would you like to be added?

Had an offline discussion with @shizhMSFT and he suggests to add two commands:
- `oras backup` (alias: `snapshot`, `pull-image`) to support saving an image from remote registry to an OCI image layout tar archive.
```console
Usage:
  oras backup --output <oci_layout_path> <reg>/<repo>[{:<tag>|@digest]}] [...]
```
- `oras restore` (alias: `push-image`) to support pushing an image from an OCI image layout tar archive to remote registry.
```console
Usage:
  oras restore --output <reg>/<repo> <oci_layout_path>[{:<tag>|@digest]}]
```

### Why is this needed for ORAS?

`oras cp` is not intuitive to use, e.g. user might use `oras push` to push images from an OCI layout and succeed with false positive (see #1353). Adding those two wrapper commands of `oras cp` can help avoid that.

### Are you willing to submit PRs to contribute to this feature?

- [ ] Yes, I am willing to implement it.

**Comment by TerryHowe** [comment]
Why not backup and restore?

**Comment by shizhMSFT** [comment]
> Why not backup and restore?

Nice input. I've revised the proposal accordingly to `oras backup` and marked `snapshot` as an alias.

**Comment by sajayantony** [comment]
Could you also include reasoning why pull doesn't work here. 

`oras pull --to-oci-layout <oci_layout_path> <reg>/<repo>[{:<tag>|@digest]}] [...]`

**Comment by sajayantony** [comment]
As per the call I'm ok if we limit to tar archive only dealing.

**Comment by shizhMSFT** [comment]
As per the call, here is an alternative without introducing new commands.
- `oras backup` --> `oras pull --archive <tar_path> <name>{:<tag>|@<digest>}`
- `oras restore` --> `oras push --archive <tar_path> <name>[:<tag>[,<tag>][...]] <file>[:<type>] [...]`

The major difference between `oras backup/restore` and `oras pull/push` is that `oras backup/restore` can take multiple images across registries and repositories, and backup/restore the entire repository as well as referrers.

Example usages are
```sh
# working with remote registry
oras pull localhost:5000/hello:v1 --archive hello.tar
oras push localhost:5000/hello:v1 --archive hello.tar

# working with multiple images
oras pull localhost:5000/hello:v1,foo,bar --archive hello.tar # backup v1, foo, and bar
oras push localhost:5000/hello:v1,foo --archive hello.tar     # restore v1 and foo, and ignore bar

# working with referrers
oras pull --include-referrers localhost:5000/hello:v1 --archive hello.tar
oras push --include-referrers localhost:5000/hello:v1 --archive hello.tar

# working with OCI layout. UX would be better if using `oras cp`.
oras pull --oci-layout layout.tar:v1 --archive hello.tar
oras pull --oci-layout layout-dir:v1 --archive hello.tar
oras push --oci-layout layout-dir:v1 --archive hello.tar
```

**Comment by qweeah** [comment]
> oras restore --> oras push --archive <tar_path> <name>[:<tag>[,<tag>][...]] <file>[:<type>] [...]

I think the usage is change to `oras push --archive <tar_path> <name>[:<tag>[,<tag>][...]]`.

**Comment by TerryHowe** [comment]
Currently, pull skips successors though, right?

**Comment by TerryHowe** [comment]
This issue is what I was requesting on https://github.com/oras-project/oras/issues/730

**Comment by shizhMSFT** [comment]
> This issue is what I was requesting on #730

Yes, exactly. The difference is that the requirements are getting clearer after 1.5 years where more tools, including docker, are adopting OCI image layout.

**Comment by qweeah** [comment]
There is a little difference between [the proposal](https://github.com/oras-project/oras/issues/1366#issuecomment-2246699831) and #730: cross-repository operation is not in scope here.

**Comment by liamwh** [comment]
Please forgive me @qweeah, I'm not sure I follow.

For my use case I just want to be able to pull and image and save it to disk, then be able to push it again to a different registry later. Is this being worked on and supported?

**Comment by qweeah** [comment]
> Please forgive me @qweeah, I'm not sure I follow.
> 
> For my use case I just want to be able to pull and image and save it to disk, then be able to push it again to a different registry later. Is this being worked on and supported?

@liamwh Yes, it is already supported. You can use `oras cp` with `--to-oci-layout`(pull) and `--from-oci-layout`(push)

```console
> # pull an image to local folder
> oras cp ghcr.io/oras-project/oras:v1.2.0 --to-oci-layout test:v1.2.0
✓ Copied  application/vnd.oci.image.config.v1+json                                                        1.56/1.56 kB 100.00%  626µs
  └─ sha256:9cd577b9700fa49ac6b2b8eca69731494a6212c7c7fca50669601e8068acc420
...
Copied [registry] ghcr.io/oras-project/oras:v1.2.0 => [oci-layout] test:v1.2.0
Digest: sha256:0087224dd0decc354b5b0689068fbbc40cd5dc3dbf65fcb3868dfbd363dc790b
```

An [OCI image layout](https://github.com/opencontainers/image-spec/blob/v1.1.0/image-layout.md#oci-image-layout-specification) folder `test` will be created and contains the pulled image tagged `v1.2.0`
```console
> tree .
.
└── test
    ├── blobs
    │   └── sha256
    │       ├── 0087224dd0decc354b5b0689068fbbc40cd5dc3dbf65fcb3868dfbd363dc790b
    ...
    ├── index.json
    ├── ingest
    └── oci-layout
```

You can push the image from the folder to your own registry like:

```console
> # push an image from local folder
> oras cp --from-oci-layout test:v1.2.0 localhost:5000/oras:v1.2.0
✓ Copied  application/vnd.oci.image.config.v1+json                                                        1.57/1.57 kB 100.00%   86ms
  └─ sha256:306409c919f25a0c7900c4610a925b53f6fb32a251c0ec81ab8b4381296b0cf3
...
  └─ sha256:0087224dd0decc354b5b0689068fbbc40cd5dc3dbf65fcb3868dfbd363dc790b
Copied [oci-layout] test:v1.2.0 => [registry] localhost:5000/oras:v1.2.0
Digest: sha256:0087224dd0decc354b5b0689068fbbc40cd5dc3dbf65fcb3868dfbd363dc790b
```

**Comment by liamwh** [comment]
Ah ok great, thank you 🙏 

Is it then correct to assume that this is also supported from the Go SDK?

**Comment by liamwh** [comment]
And is it possible to save the file as a tar instead of unpacking it?

**Comment by qweeah** [comment]
Saving(pulling) to a tar file is not supported, but reading(pushing) from a tar file is supported.

**Comment by liamwh** [comment]
Thanks for clarifying. Is it possible to save to a tar file instead of unpacking the contents to the file system?

**Comment by TerryHowe** [comment]
>  Is it possible to save to a tar file instead of unpacking the contents to the file system?

No

**Comment by liamwh** [comment]
Please forgive my ignorance, I'm just trying to understand: why not skip the extraction and support just downloading the tar, given that's what an OCI image actually is?

**Comment by TerryHowe** [comment]
> Please forgive my ignorance, I'm just trying to understand: why not skip the extraction and support just downloading the tar, given that's what an OCI image actually is?

My understanding is that is exactly what this issue is about.

**Comment by qweeah** [comment]
> why not skip the extraction and support just downloading the tar, given that's what an OCI image actually is?

@liamwh Image layer files can be tarball and oras will not extract layer files. 

However, the image itself, consisting of layer tarball files and other metadata, is not a tarball. The OCI layout folder won't be compressed by oras.

**Comment by Wwwsylvia** [comment]
Reopening this issue as the implementation is not yet complete.
